M2/Macaulay2/m2/enginering.m2

```diff
@@ -225,7 +225,40 @@ commonEngineRingInitializations = (F) -> (
      *-
      )
 
------------------------------------------------------------------------------
+------------------------------------------------------------------------------
+-- TODO improve this, or deprecate promote/lift(List,R,S)
+defaultDegMap := (R,S) -> (
+    n:=degreeLength S-degreeLength R;
+    if n==0 then identity
+    else if n<0 then d -> take(d,degreeLength S)
+    else d -> d|toList(n:0)
+    )
+
+-- automate promote
+setupPromote = method()
+setupPromote (Function,Ring,Ring,Function) := (f,R,S,degmap) -> (
+    promote(R,S) := (a,S) -> f a;
+    promote(List,R,S) := (m,R,S) -> apply(m,degmap);
+    -- promote(Matrix,R,S) :=
+    -- promote(MutableMatrix,R,S) := -- doesn't work, cf https://github.com/Macaulay2/M2/issues/2192
```

Comment:
I'm confused about these commented lines. Don't you have these exact methods implemented two lines below?

Reply:
yes, but not as `S ** M`

Reply:
Why would it be? That wouldn't even work because `Ring ** Matrix` calls `Ring ** Module` which then calls `promote(Matrix, Ring)`, so it would be a loop.

Reply:
I think the comment (which I wrote a while ago) only applies to the `MutableMatrix`. I'll just remove these lines.